Sources/GRPCHTTP2Core/GRPCStreamStateMachine.swift

```diff
@@ -0,0 +1,1422 @@
+    outboundEncoding: CompressionAlgorithm?,
+    acceptedEncodings: [CompressionAlgorithm],
+    customMetadata: Metadata
+  ) -> HPACKHeaders {
+    var headers = HPACKHeaders()
+    headers.reserveCapacity(7 + customMetadata.count)
+
+    // Add required headers
+    // See https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md#requests
+    headers.path = methodDescriptor
+    headers.scheme = scheme
+    headers.method = "POST"
+    headers.contentType = .protobuf
+    headers.te = "trailers"  // Used to detect incompatible proxies
+
+    if let encoding = outboundEncoding {
+      headers.encoding = encoding
+    }
+
+    if !acceptedEncodings.isEmpty {
+      headers.acceptedEncodings = acceptedEncodings
+    }
+
+    for metadataPair in customMetadata {
+      headers.add(name: metadataPair.key, value: metadataPair.value.encoded())
+    }
+
+    return headers
+  }
+
+  private mutating func clientSend(
+    metadata: Metadata,
+    configuration: GRPCStreamStateMachineConfiguration.ClientConfiguration
+  ) throws -> HPACKHeaders {
+    // Client sends metadata only when opening the stream.
+    switch self.state {
+    case .clientIdleServerIdle(let state):
+      self.state = .clientOpenServerIdle(
+        .init(
+          previousState: state,
+          compressionAlgorithm: configuration.outboundEncoding,
+          decompressionConfiguration: .decompressionNotYetKnown
+        )
+      )
+      return self.makeClientHeaders(
+        methodDescriptor: configuration.methodDescriptor,
+        scheme: configuration.scheme,
+        outboundEncoding: configuration.outboundEncoding,
+        acceptedEncodings: configuration.acceptedEncodings,
+        customMetadata: metadata
+      )
+    case .clientOpenServerIdle, .clientOpenServerOpen, .clientOpenServerClosed:
+      throw self.assertionFailureAndCreateRPCErrorOnInternalError(
+        "Client is already open: shouldn't be sending metadata."
+      )
+    case .clientClosedServerIdle, .clientClosedServerOpen, .clientClosedServerClosed:
+      throw self.assertionFailureAndCreateRPCErrorOnInternalError(
+        "Client is closed: can't send metadata."
+      )
+    }
+  }
+
+  private mutating func clientSend(message: [UInt8], endStream: Bool) throws {
+    // Client sends message.
+    switch self.state {
+    case .clientIdleServerIdle:
+      throw self.assertionFailureAndCreateRPCErrorOnInternalError("Client not yet open.")
+    case .clientOpenServerIdle(var state):
+      state.framer.append(message)
+      if endStream {
+        self.state = .clientClosedServerIdle(.init(previousState: state))
+      } else {
+        self.state = .clientOpenServerIdle(state)
+      }
+    case .clientOpenServerOpen(var state):
+      state.framer.append(message)
+      if endStream {
+        self.state = .clientClosedServerOpen(.init(previousState: state))
+      } else {
+        self.state = .clientOpenServerOpen(state)
+      }
+    case .clientOpenServerClosed(let state):
```

Comment:
This one seems interesting to me. Shouldn't the `writer.write` method throw an error in this case and the write shouldn't even make it here or is there a potential race?